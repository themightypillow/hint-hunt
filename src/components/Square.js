import React from "react";
import { observer } from "mobx-react";

@observer
class Square extends React.Component {
  handleMouseDown = () => {
    this.props.board.setInitial(this.props.letter);
  }

  handleMouseOver = () => {
    this.props.board.startChain(this.props.letter);
  }
  
  render() {
    return (
      <div 
        className="square"
        onMouseDown={this.handleMouseDown}
        onMouseOver={this.handleMouseOver}
      >
        {
          this.props.letter.classes.map(
            (name, index) => <div key={`${this.props.letter.letter}${index}`} className={name}></div>
          )
        }
        {this.props.letter.letter}
      </div>
    );
  }
}

export default Square;