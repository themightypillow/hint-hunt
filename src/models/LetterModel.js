import { configure, observable, action } from "mobx";

// state can only be changed inside @action methods
configure({ enforceActions: "observed" });

class LetterModel {
  @observable letter;
  @observable classes = [];
  @observable index;
  columns = 13;

  constructor(letter, index) {
    this.letter = letter;
    this.index = index;
  }

  coords() {
    return [Math.floor(this.index / this.columns), this.index % this.columns];
  }

  @action pushClass = (name) => {
    this.classes.push(name);
  }

  @action popClass = () => {
    this.classes.pop();
  }
}

export default LetterModel;