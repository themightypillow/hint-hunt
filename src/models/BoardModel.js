import { configure, action, observable } from "mobx";
import LetterModel from "./LetterModel";

// state can only be changed inside @action methods
configure({ enforceActions: "observed" });

class BoardModel {
  @observable grid = [];
  @observable guess = [];
  @observable initial;
  @observable hover;
  @observable isMouseDown = false;
  columns = 13;

  constructor(grid) {
    this.grid = grid.map((letter, index) => new LetterModel(letter, index));
  }

  getIndex(row, column) {
    return row * this.columns + column;
  }

  @action startGuess = (letter) => {
    this.isMouseDown = true;
    this.initial = letter;
  }

  @action startChain = (letter) => {
    if(this.isMouseDown) {
      this.hover = letter;
      this.popChainClasses();
      this.guess.length = 0;
      const [row1, col1] = this.initial.coords();
      const [row2, col2] = letter.coords();
      if(row1 === row2 && col1 === col2) { // hovering over initial letter
        this.makeSingleChain();
      }
      else if(row1 === row2) { // horizontal
        this.makeChain(row1, col1, col2, false);
      }
      else if(col1 === col2) { // vertical
        this.makeChain(col1, row1, row2, true);
      }
      else if((row2 - row1) === (col2 - col1)) { // backward diagonal
        this.makeDiagonalChain(row1, row2, false);
      }
      else if((row2 - row1) === -(col2 - col1)) { // forward diagonal
        this.makeDiagonalChain(row1, row2, true);
      }
      else { // chain not possible
        this.makeSingleChain();
      }
      console.log(this.guess.map(char => char.letter));
    }
  }

  @action makeSingleChain = () => {
    this.guess.push(this.initial);
    this.initial.pushClass("square_circle");
  }

  @action makeChain = (same, diff1, diff2, vertical) => {
    let first = diff1 < diff2 ? this.initial : this.hover;
    let last = diff1 < diff2 ? this.hover : this.initial;
    this.guess.push(first);
    first.pushClass(vertical ? "square_top" : "square_left");
    for(let i = Math.min(diff1, diff2) + 1; i < Math.max(diff1, diff2); i++) {
      const coords = vertical ? [i, same] : [same, i];
      const next = this.grid[this.getIndex(...coords)];
      this.guess.push(next);
      next.pushClass(vertical ? "square_topBottom" : "square_leftRight");
    }
    this.guess.push(last);
    last.pushClass(vertical ? "square_bottom" : "square_right");
  }

  @action makeDiagonalChain = (row1, row2, forward) => {
    let first = row1 < row2 ? this.initial : this.hover;
    let last = row1 < row2 ? this.hover : this.initial;
    this.guess.push(first);
    first.pushClass(forward ? "square_topRight" : "square_topLeft");
    let firstCoords = first.coords();
    for(let i = 1; i < Math.max(row1, row2) - Math.min(row1, row2); i++) {
      const next = this.grid[
        this.getIndex(firstCoords[0] + i, forward ? firstCoords[1] - i : firstCoords[1] + i)
      ];
      this.guess.push(next);
      next.pushClass(forward ? "square_forwardSlash" : "square_backSlash");
    }
    this.guess.push(last);
    last.pushClass(forward ? "square_bottomLeft" : "square_bottomRight");
  }

  popChainClasses = () => {
    for(let i = 0; i < this.guess.length; i++) {
      this.guess[i].popClass();
    }
  }

  @action mouseUp = () => {
    this.isMouseDown = false;
    this.popChainClasses();
    this.guess.length = 0;
  }
}

export default BoardModel;