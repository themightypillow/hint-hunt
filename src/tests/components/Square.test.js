import React from "react";
import { shallow } from "enzyme";
import Square from "../../components/Square";
import LetterModel from "../../models/LetterModel";
import BoardModel from "../../models/BoardModel";

test("should render Square with no classes", () => {
  const letter = new LetterModel("A", 0);
  const wrapper = shallow(<Square letter={letter} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render Square with one class", () => {
  const letter = new LetterModel("A", 0);
  letter.pushClass("one");
  const wrapper = shallow(<Square letter={letter} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render Square with two classes", () => {
  const letter = new LetterModel("A", 0);
  letter.pushClass("one");
  letter.pushClass("two");
  const wrapper = shallow(<Square letter={letter} />);
  expect(wrapper).toMatchSnapshot();
});

describe("mouse events", () => {
  const letter = new LetterModel("A", 0);
  const board = new BoardModel(["A", "B"], ["A"]);
  const wrapper = shallow(<Square board={board} letter={letter} />);

  test("should run function on mousedown event", () => {
    wrapper.find(".square").simulate("mousedown");
    expect(board.isMouseDown).toBeTruthy();
  });

  test("should run function on mouseover event", () => {
    wrapper.find(".square").simulate("mouseover");
    expect(board.initial).toEqual(board.hover);
  });
});