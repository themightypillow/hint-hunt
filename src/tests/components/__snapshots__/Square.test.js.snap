// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Square with no classes 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Square
    letter={
      LetterModel {
        "classes": Array [],
        "columns": 13,
        "index": 0,
        "letter": "A",
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "LetterModel@6.keys",
            "observers": Set {},
          },
          "name": "LetterModel@6",
          "target": [Circular],
          "values": Map {
            "letter" => "A",
            "classes" => Array [],
            "index" => 0,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [],
        "A",
      ],
      "className": "square",
      "onMouseDown": [Function],
      "onMouseOver": [Function],
    },
    "ref": null,
    "rendered": Array [
      "A",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [],
          "A",
        ],
        "className": "square",
        "onMouseDown": [Function],
        "onMouseOver": [Function],
      },
      "ref": null,
      "rendered": Array [
        "A",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render Square with one class 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Square
    letter={
      LetterModel {
        "classes": Array [
          "one",
        ],
        "columns": 13,
        "index": 0,
        "letter": "A",
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "LetterModel@8.keys",
            "observers": Set {},
          },
          "name": "LetterModel@8",
          "target": [Circular],
          "values": Map {
            "letter" => "A",
            "classes" => Array [
              "one",
            ],
            "index" => 0,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <div
            className="one"
          />,
        ],
        "A",
      ],
      "className": "square",
      "onMouseDown": [Function],
      "onMouseOver": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "A0",
        "nodeType": "host",
        "props": Object {
          "className": "one",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "A",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div
              className="one"
            />,
          ],
          "A",
        ],
        "className": "square",
        "onMouseDown": [Function],
        "onMouseOver": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "A0",
          "nodeType": "host",
          "props": Object {
            "className": "one",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "A",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render Square with two classes 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Square
    letter={
      LetterModel {
        "classes": Array [
          "one",
          "two",
        ],
        "columns": 13,
        "index": 0,
        "letter": "A",
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "LetterModel@10.keys",
            "observers": Set {},
          },
          "name": "LetterModel@10",
          "target": [Circular],
          "values": Map {
            "letter" => "A",
            "classes" => Array [
              "one",
              "two",
            ],
            "index" => 0,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <div
            className="one"
          />,
          <div
            className="two"
          />,
        ],
        "A",
      ],
      "className": "square",
      "onMouseDown": [Function],
      "onMouseOver": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "A0",
        "nodeType": "host",
        "props": Object {
          "className": "one",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "A1",
        "nodeType": "host",
        "props": Object {
          "className": "two",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "A",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div
              className="one"
            />,
            <div
              className="two"
            />,
          ],
          "A",
        ],
        "className": "square",
        "onMouseDown": [Function],
        "onMouseOver": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "A0",
          "nodeType": "host",
          "props": Object {
            "className": "one",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "A1",
          "nodeType": "host",
          "props": Object {
            "className": "two",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "A",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
